[
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "I think the problem with #23799 was that it introduced a second outer loop `while i < test_count:` which broke the `--failfast` option. If that option was chosen and a test failed, the break would leave the inner loop, but since the condition of the outer loop was not fulfilled, the inner loop would just be started again. \r\n\r\nAs far as I can see, this problem is still there (even though I can't reproduce the \"pop from empty list\" with this branch).",
      "created_at" : "2022-01-07T14:12:32Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1007436418",
      "id" : 1007436418,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII5848DEKC",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1007436418/reactions"
      },
      "updated_at" : "2022-01-07T14:14:25Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1007436418",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/48763452?v=4",
         "events_url" : "https://api.github.com/users/mzumsande/events{/privacy}",
         "followers_url" : "https://api.github.com/users/mzumsande/followers",
         "following_url" : "https://api.github.com/users/mzumsande/following{/other_user}",
         "gists_url" : "https://api.github.com/users/mzumsande/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/mzumsande",
         "id" : 48763452,
         "login" : "mzumsande",
         "node_id" : "MDQ6VXNlcjQ4NzYzNDUy",
         "organizations_url" : "https://api.github.com/users/mzumsande/orgs",
         "received_events_url" : "https://api.github.com/users/mzumsande/received_events",
         "repos_url" : "https://api.github.com/users/mzumsande/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/mzumsande/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/mzumsande/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/mzumsande"
      }
   },
   {
      "author_association" : "NONE",
      "body" : "I see the same behavior as @mzumsande: `--failfast` option doesn't work, and can't reproduce \"pop from empty list\".\r\n\r\nHere is a suggested fix. (I don't know how to suggest changes on lines unchanged.) Any feedback is appreciated as it is a learning opportunity.\r\n\r\n- add `failed` attribute to `TestHandler` class\r\n\r\n```python\r\nclass TestHandler:\r\n    \"\"\"\r\n    Trigger the test scripts passed in via the list.\r\n    \"\"\"\r\n\r\n    def __init__(self, *, num_tests_parallel, tests_dir, tmpdir, test_list, flags, use_term_control, failed):\r\n        assert num_tests_parallel >= 1\r\n        self.num_jobs = num_tests_parallel\r\n        self.tests_dir = tests_dir\r\n        self.tmpdir = tmpdir\r\n        self.test_list = test_list\r\n        self.flags = flags\r\n        self.jobs = []\r\n        self.use_term_control = use_term_control\r\n        self.failed = failed\r\n\r\njob_queue = TestHandler(\r\n    num_tests_parallel=jobs,\r\n    tests_dir=tests_dir,\r\n    tmpdir=tmpdir,\r\n    test_list=test_list,\r\n    flags=flags,\r\n    use_term_control=use_term_control,\r\n    failed=False\r\n)\r\n```\r\n\r\n- modify 'failed' in case of fast fail\r\n\r\n```python\r\nif failfast:\r\n    job_queue.failed = True\r\n    logging.debug(\"Early exiting after test failure\")\r\n    break\r\n```\r\n\r\n- modify `done` method\r\n\r\n```python\r\ndef done(self):\r\n    return self.failed or (not (self.jobs or self.test_list))\r\n```\r\n\r\n",
      "created_at" : "2022-01-13T02:15:46Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1011681271",
      "id" : 1011681271,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII5848TQf3",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1011681271/reactions"
      },
      "updated_at" : "2022-01-13T02:15:46Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1011681271",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/91135419?v=4",
         "events_url" : "https://api.github.com/users/pg156/events{/privacy}",
         "followers_url" : "https://api.github.com/users/pg156/followers",
         "following_url" : "https://api.github.com/users/pg156/following{/other_user}",
         "gists_url" : "https://api.github.com/users/pg156/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/pg156",
         "id" : 91135419,
         "login" : "pg156",
         "node_id" : "MDQ6VXNlcjkxMTM1NDE5",
         "organizations_url" : "https://api.github.com/users/pg156/orgs",
         "received_events_url" : "https://api.github.com/users/pg156/received_events",
         "repos_url" : "https://api.github.com/users/pg156/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/pg156/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/pg156/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/pg156"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "> I think the problem with #23799 was that it introduced a second outer loop `while i < test_count:` which broke the `--failfast` option. If that option was chosen and a test failed, the break would leave the inner loop, but since the condition of the outer loop was not fulfilled, the inner loop would just be started again.\r\n\r\nfailfast is very useful when having smol computing power\r\n\r\n~~if this languge supports multiple block breaking (as many\r\nothers do) I suppose why wouldnt you have it break 2\r\nas an _urgent_ amend to 23799 while something more\r\ncalmly thought is coming~~\r\n\r\nTODO: add some tests to test the tests framework\r\n",
      "created_at" : "2022-01-21T22:14:13Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1018903017",
      "id" : 1018903017,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII5848uznp",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1018903017/reactions"
      },
      "updated_at" : "2022-01-21T22:34:32Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1018903017",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/52637275?v=4",
         "events_url" : "https://api.github.com/users/katesalazar/events{/privacy}",
         "followers_url" : "https://api.github.com/users/katesalazar/followers",
         "following_url" : "https://api.github.com/users/katesalazar/following{/other_user}",
         "gists_url" : "https://api.github.com/users/katesalazar/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/katesalazar",
         "id" : 52637275,
         "login" : "katesalazar",
         "node_id" : "MDQ6VXNlcjUyNjM3Mjc1",
         "organizations_url" : "https://api.github.com/users/katesalazar/orgs",
         "received_events_url" : "https://api.github.com/users/katesalazar/received_events",
         "repos_url" : "https://api.github.com/users/katesalazar/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/katesalazar/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/katesalazar/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/katesalazar"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "<!--e57a25ab6845829454e8d69fc972939a-->\n\nThe following sections might be updated with supplementary metadata relevant to reviewers and maintainers.\n\n<!--174a7506f384e20aa4161008e828411d-->\n### Conflicts\nReviewers, this pull request conflicts with the following ones:\n\n* [#24195](https://github.com/bitcoin/bitcoin/pull/24195) (test: Fix failfast option for functional test runner by mzumsande)\n\nIf you consider this pull request important, please also help to review the conflicting pull requests. Ideally, start with the one that should be merged first.",
      "created_at" : "2022-01-28T18:30:56Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1024498196",
      "id" : 1024498196,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII5849EJoU",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1024498196/reactions"
      },
      "updated_at" : "2022-01-28T18:30:56Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1024498196",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/39886733?v=4",
         "events_url" : "https://api.github.com/users/DrahtBot/events{/privacy}",
         "followers_url" : "https://api.github.com/users/DrahtBot/followers",
         "following_url" : "https://api.github.com/users/DrahtBot/following{/other_user}",
         "gists_url" : "https://api.github.com/users/DrahtBot/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/DrahtBot",
         "id" : 39886733,
         "login" : "DrahtBot",
         "node_id" : "MDQ6VXNlcjM5ODg2NzMz",
         "organizations_url" : "https://api.github.com/users/DrahtBot/orgs",
         "received_events_url" : "https://api.github.com/users/DrahtBot/received_events",
         "repos_url" : "https://api.github.com/users/DrahtBot/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/DrahtBot/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/DrahtBot/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/DrahtBot"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "<!--cf906140f33d8803c4a75a2196329ecb-->\nð This pull request conflicts with the target branch and [needs rebase](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#rebasing-changes).\n\n<sub>Want to unsubscribe from rebase notifications on this pull request? Just convert this pull request to a \"draft\".</sub>",
      "created_at" : "2022-02-07T16:32:54Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1031668248",
      "id" : 1031668248,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII5849fgIY",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1031668248/reactions"
      },
      "updated_at" : "2022-02-07T16:32:54Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1031668248",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/39886733?v=4",
         "events_url" : "https://api.github.com/users/DrahtBot/events{/privacy}",
         "followers_url" : "https://api.github.com/users/DrahtBot/followers",
         "following_url" : "https://api.github.com/users/DrahtBot/following{/other_user}",
         "gists_url" : "https://api.github.com/users/DrahtBot/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/DrahtBot",
         "id" : 39886733,
         "login" : "DrahtBot",
         "node_id" : "MDQ6VXNlcjM5ODg2NzMz",
         "organizations_url" : "https://api.github.com/users/DrahtBot/orgs",
         "received_events_url" : "https://api.github.com/users/DrahtBot/received_events",
         "repos_url" : "https://api.github.com/users/DrahtBot/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/DrahtBot/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/DrahtBot/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/DrahtBot"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "lgtm, but this still needs rebase, either here or in a separate pull.",
      "created_at" : "2022-03-22T10:37:40Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1075006648",
      "id" : 1075006648,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII585AE0y4",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1075006648/reactions"
      },
      "updated_at" : "2022-03-22T10:37:40Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1075006648",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/6399679?v=4",
         "events_url" : "https://api.github.com/users/MarcoFalke/events{/privacy}",
         "followers_url" : "https://api.github.com/users/MarcoFalke/followers",
         "following_url" : "https://api.github.com/users/MarcoFalke/following{/other_user}",
         "gists_url" : "https://api.github.com/users/MarcoFalke/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/MarcoFalke",
         "id" : 6399679,
         "login" : "MarcoFalke",
         "node_id" : "MDQ6VXNlcjYzOTk2Nzk=",
         "organizations_url" : "https://api.github.com/users/MarcoFalke/orgs",
         "received_events_url" : "https://api.github.com/users/MarcoFalke/received_events",
         "repos_url" : "https://api.github.com/users/MarcoFalke/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/MarcoFalke/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/MarcoFalke/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/MarcoFalke"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "Concept ACK",
      "created_at" : "2022-04-05T14:08:13Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1088751221",
      "id" : 1088751221,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII585A5QZ1",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1088751221/reactions"
      },
      "updated_at" : "2022-04-05T14:08:13Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1088751221",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/126646?v=4",
         "events_url" : "https://api.github.com/users/laanwj/events{/privacy}",
         "followers_url" : "https://api.github.com/users/laanwj/followers",
         "following_url" : "https://api.github.com/users/laanwj/following{/other_user}",
         "gists_url" : "https://api.github.com/users/laanwj/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/laanwj",
         "id" : 126646,
         "login" : "laanwj",
         "node_id" : "MDQ6VXNlcjEyNjY0Ng==",
         "organizations_url" : "https://api.github.com/users/laanwj/orgs",
         "received_events_url" : "https://api.github.com/users/laanwj/received_events",
         "repos_url" : "https://api.github.com/users/laanwj/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/laanwj/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/laanwj/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/laanwj"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "Closing. #23799 is fixed through #24195, and I don't feel like rebasing the other changes here.",
      "created_at" : "2022-04-05T18:43:26Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1089176490",
      "id" : 1089176490,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII585A64Oq",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1089176490/reactions"
      },
      "updated_at" : "2022-04-05T18:43:26Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1089176490",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/548488?v=4",
         "events_url" : "https://api.github.com/users/sipa/events{/privacy}",
         "followers_url" : "https://api.github.com/users/sipa/followers",
         "following_url" : "https://api.github.com/users/sipa/following{/other_user}",
         "gists_url" : "https://api.github.com/users/sipa/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/sipa",
         "id" : 548488,
         "login" : "sipa",
         "node_id" : "MDQ6VXNlcjU0ODQ4OA==",
         "organizations_url" : "https://api.github.com/users/sipa/orgs",
         "received_events_url" : "https://api.github.com/users/sipa/received_events",
         "repos_url" : "https://api.github.com/users/sipa/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/sipa/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/sipa/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/sipa"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "Marked up for grabs.",
      "created_at" : "2022-04-05T18:56:06Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/23995#issuecomment-1089191043",
      "id" : 1089191043,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/23995",
      "node_id" : "IC_kwDOABII585A67yD",
      "performed_via_github_app" : null,
      "reactions" : {
         "+1" : 0,
         "-1" : 0,
         "confused" : 0,
         "eyes" : 0,
         "heart" : 0,
         "hooray" : 0,
         "laugh" : 0,
         "rocket" : 0,
         "total_count" : 0,
         "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1089191043/reactions"
      },
      "updated_at" : "2022-04-05T18:56:06Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/1089191043",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/548488?v=4",
         "events_url" : "https://api.github.com/users/sipa/events{/privacy}",
         "followers_url" : "https://api.github.com/users/sipa/followers",
         "following_url" : "https://api.github.com/users/sipa/following{/other_user}",
         "gists_url" : "https://api.github.com/users/sipa/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/sipa",
         "id" : 548488,
         "login" : "sipa",
         "node_id" : "MDQ6VXNlcjU0ODQ4OA==",
         "organizations_url" : "https://api.github.com/users/sipa/orgs",
         "received_events_url" : "https://api.github.com/users/sipa/received_events",
         "repos_url" : "https://api.github.com/users/sipa/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/sipa/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/sipa/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/sipa"
      }
   }
]
